This project investigates the factors influencing academic perfomance among students, it looks into factors such as: attendance, parents education, internet access as well as extra curricular activities.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [4]:
df = pd.read_csv('data/Student_Performance.csv')

In [5]:
df.head()

,student_id,age,gender,school_type,parent_education,study_hours,attendance_percentage,internet_access,travel_time,extra_activities,study_method,math_score,science_score,english_score,overall_score,final_grade
0,1,14,male,public,post graduate,3.1,84.3,yes,<15 min,yes,notes,42.7,55.4,57.0,53.1,e
1,2,18,female,public,graduate,3.7,87.8,yes,>60 min,no,textbook,57.6,68.8,64.8,61.3,d
2,3,17,female,private,post graduate,7.9,65.5,no,<15 min,no,notes,84.8,95.0,79.2,89.6,b
3,4,16,other,public,high school,1.1,58.1,no,15-30 min,no,notes,44.4,27.5,54.7,41.6,e
4,5,16,female,public,high school,1.3,61.0,yes,30-60 min,yes,group study,8.9,32.7,30.0,25.4,f


In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 25000 entries, 0 to 24999
Data columns (total 16 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   student_id             25000 non-null  int64  
 1   age                    25000 non-null  int64  
 2   gender                 25000 non-null  str    
 3   school_type            25000 non-null  str    
 4   parent_education       25000 non-null  str    
 5   study_hours            25000 non-null  float64
 6   attendance_percentage  25000 non-null  float64
 7   internet_access        25000 non-null  str    
 8   travel_time            25000 non-null  str    
 9   extra_activities       25000 non-null  str    
 10  study_method           25000 non-null  str    
 11  math_score             25000 non-null  float64
 12  science_score          25000 non-null  float64
 13  english_score          25000 non-null  float64
 14  overall_score          25000 non-null  float64
 15  final_grade  

In [7]:
df.describe()

,student_id,age,study_hours,attendance_percentage,math_score,science_score,english_score,overall_score
count,25000.00000,25000.000000,25000.000000,25000.000000,25000.000000,25000.000000,25000.000000,25000.000000
mean,7493.04380,16.482760,4.253224,75.084084,63.785944,63.745320,63.681948,64.006172
std,4323.56215,1.703895,2.167541,14.373171,20.875262,20.970529,20.792693,18.932025
min,1.00000,14.000000,0.500000,50.000000,0.000000,0.000000,0.000000,14.500000
25%,3743.75000,15.000000,2.400000,62.800000,48.300000,48.200000,48.300000,49.000000
50%,7461.50000,16.000000,4.300000,75.100000,64.100000,64.100000,64.200000,64.200000
75%,11252.00000,18.000000,6.100000,87.500000,80.000000,80.000000,80.000000,79.000000
max,15000.00000,19.000000,8.000000,100.000000,100.000000,100.000000,100.000000,100.000000


In [8]:
df.isnull().sum()

student_id               0
age                      0
gender                   0
school_type              0
parent_education         0
study_hours              0
attendance_percentage    0
internet_access          0
travel_time              0
extra_activities         0
study_method             0
math_score               0
science_score            0
english_score            0
overall_score            0
final_grade              0
dtype: int64

In [10]:
# 1. Average Overall Score
avg_overall = df['overall_score'].mean()

# 2. Distribution of Final Grade
grade_counts = df['final_grade'].value_counts()

# 3. Correlation between Math, Science, and English
# This checks if students who do well in one subject do well in others
subject_corr = df[['math_score', 'science_score', 'english_score']].corr()

# 4. School Type Comparison
school_dist = df['school_type'].value_counts()

# 5. Average Study Hours
avg_study = df['study_hours'].mean()

# Output results
print(f"Avg Score: {avg_overall:.2f}")
print(f"Avg Study Hours: {avg_study:.2f}")
print("\nGrade Distribution:\n", grade_counts)
print("\nCorrelation Matrix:\n", subject_corr)

Avg Score: 64.01
Avg Study Hours: 4.25

Grade Distribution:
 final_grade
d    6311
c    6161
e    5672
f    2955
b    2696
a    1205
Name: count, dtype: int64

Correlation Matrix:
                math_score  science_score  english_score
math_score       1.000000       0.784111       0.786315
science_score    0.784111       1.000000       0.783951
english_score    0.786315       0.783951       1.000000


Average Performance: The mean overall_score for the 25,000 students is 64.01.
Study Habits: On average, students commit 4.25 study_hours per week.
Core Subject Synergy: There is a strong positive correlation (approx. 0.78) between math_score, science_score, and english_score. This suggests that students who excel in one core area are highly likely to excel in the others.

In [11]:
# Do males and females differ in overall_score?
print(df.groupby('gender')['overall_score'].mean())

# Does one gender perform better in math vs English?
print(df.groupby('gender')[['math_score', 'english_score']].mean())

# Does study_hours differ by gender?
print(df.groupby('gender')['study_hours'].mean())

gender
female    64.266828
male      64.072645
other     63.686069
Name: overall_score, dtype: float64
        math_score  english_score
gender                           
female   64.045428      63.893317
male     63.872754      63.661150
other    63.447170      63.495167
gender
female    4.271930
male      4.258021
other     4.230226
Name: study_hours, dtype: float64


In [ ]:
# Do private schools outperform public schools?
print(df.groupby('school_type')['overall_score'].mean())

# Is attendance higher in one school type?
print(df.groupby('school_type')['attendance_percentage'].mean())

# Is internet access more common in one group?
print(pd.crosstab(df['school_type'], df['internet_access'], normalize='index'))

school_type
private    63.985226
public     64.027886
Name: overall_score, dtype: float64
school_type
private    75.148149
public     75.017670
Name: attendance_percentage, dtype: float64
internet_access        no       yes
school_type                        
private          0.150727  0.849273
public           0.151120  0.848880


In [14]:
# Does parent_education impact overall_score?
print(df.groupby('parent_education')['overall_score'].mean())

# Does it affect study_hours?
print(df.groupby('parent_education')['study_hours'].mean())

# Does it affect final_grade distribution?
print(pd.crosstab(df['parent_education'], df['final_grade'], normalize='index'))

parent_education
diploma          64.651484
graduate         63.956991
high school      63.386492
no formal        63.907085
phd              63.538637
post graduate    64.562917
Name: overall_score, dtype: float64
parent_education
diploma          4.338781
graduate         4.231815
high school      4.177455
no formal        4.230792
phd              4.241432
post graduate    4.295520
Name: study_hours, dtype: float64
final_grade              a         b         c         d         e         f
parent_education                                                            
diploma           0.047520  0.117756  0.257534  0.247102  0.215345  0.114743
graduate          0.045069  0.109280  0.259511  0.242064  0.222922  0.121153
high school       0.045660  0.099405  0.246136  0.246136  0.246136  0.116528
no formal         0.054180  0.105663  0.235597  0.254229  0.227016  0.123315
phd               0.047806  0.100270  0.233145  0.269429  0.228487  0.120863
post graduate     0.049094  0.114156  0

Female students generally outperform males in overall_score, math_score, and english_score, though this advantage is closely mirrored by their higher average study_hours. Similarly, students whose parents hold a diploma, post graduate or have no formal education achieve the highest scores, a trend that again correlates directly with their increased time spent studying. Ultimately, the data suggests that study_hours is the primary driver of academic success across all demographic groups, outweighing factors like gender or parental background.

In [17]:
# correlation between hours and scores
correlation = df['study_hours'].corr(df['overall_score'])
print(f"Correlation between Study Hours and Overall Score: {correlation:.4f}")

# Check for Diminishing Returns
# Rounding hours to the nearest whole number to see the trend more clearly
df['hours_rounded'] = df['study_hours'].round()
diminishing_returns = df.groupby('hours_rounded')['overall_score'].mean()
print("\nScore by Study Hours (Trend):")
print(diminishing_returns)

# Effectiveness by Study Method
study_method_perf = df.groupby('study_method')['overall_score'].mean()
print("\nPerformance by Study Method:")
print(study_method_perf)

# Hours spent on each study method
study_method_hours = df.groupby('study_method')['study_hours'].mean()
print(study_method_hours)

Correlation between Study Hours and Overall Score: 0.9058

Score by Study Hours (Trend):
hours_rounded
0.0    34.649405
1.0    38.013770
2.0    46.277071
3.0    54.084270
4.0    62.327444
5.0    69.967358
6.0    78.167452
7.0    85.334922
8.0    91.323884
Name: overall_score, dtype: float64

Performance by Study Method:
study_method
coaching         64.368405
group study      63.249487
mixed            63.613960
notes            63.895318
online videos    64.686905
textbook         64.238122
Name: overall_score, dtype: float64
study_method
coaching         4.256831
group study      4.176919
mixed            4.237941
notes            4.264562
online videos    4.321503
textbook         4.261264
Name: study_hours, dtype: float64


In [18]:
# Attendance vs. Study Hours Predictability
# Comparing which has a higher correlation with overall_score
attendance_corr = df['attendance_percentage'].corr(df['overall_score'])
study_corr = df['study_hours'].corr(df['overall_score'])
print(f"Correlation (Attendance vs Score): {attendance_corr:.4f}")
print(f"Correlation (Study Hours vs Score): {study_corr:.4f}")

# What happens below 75% attendance?
low_attendance_avg = df[df['attendance_percentage'] < 75]['overall_score'].mean()
high_attendance_avg = df[df['attendance_percentage'] >= 75]['overall_score'].mean()
print(f"\nAvg Score (Attendance < 75%): {low_attendance_avg:.2f}")
print(f"Avg Score (Attendance >= 75%): {high_attendance_avg:.2f}")


Correlation (Attendance vs Score): 0.2928
Correlation (Study Hours vs Score): 0.9058

Avg Score (Attendance < 75%): 59.14
Avg Score (Attendance >= 75%): 68.83


In [19]:
# Checking if internet access correlates with higher scores or acts as a distraction
internet_impact = df.groupby('internet_access')['overall_score'].mean()
print("\nImpact of Internet Access on Scores:")
print(internet_impact)


Impact of Internet Access on Scores:
internet_access
no     63.774821
yes    64.047294
Name: overall_score, dtype: float64


In [20]:
# Checking if long commutes "steal" time from studying or reduce attendance
travel_impact = df.groupby('travel_time')[['overall_score', 'study_hours', 'attendance_percentage']].mean()
print("\nImpact of Travel Time:")
print(travel_impact)


Impact of Travel Time:
             overall_score  study_hours  attendance_percentage
travel_time                                                   
15-30 min        63.986136     4.240286              75.017070
30-60 min        63.725185     4.239117              75.101797
<15 min          64.246695     4.288428              75.144181
>60 min          64.076987     4.246108              75.075284
